csharp/src/Google.Protobuf/CodedOutputWriter.cs

```diff
@@ -0,0 +1,666 @@
+        /// <param name="b2">The second byte of the encoded tag</param>
+        public void WriteRawTag(byte b1, byte b2)
+        {
+            if (buffered + 1 > span.Length)
+            {
+                EnsureMore(2);
+            }
+
+            span[buffered++] = b1;
+            span[buffered++] = b2;
+        }
+
+        /// <summary>
+        /// Writes the given three-byte tag directly to the stream.
+        /// </summary>
+        /// <param name="b1">The first byte of the encoded tag</param>
+        /// <param name="b2">The second byte of the encoded tag</param>
+        /// <param name="b3">The third byte of the encoded tag</param>
+        public void WriteRawTag(byte b1, byte b2, byte b3)
+        {
+            if (buffered + 2 > span.Length)
+            {
+                EnsureMore(3);
+            }
+
+            span[buffered++] = b1;
+            span[buffered++] = b2;
+            span[buffered++] = b3;
+        }
+
+        /// <summary>
+        /// Writes the given four-byte tag directly to the stream.
+        /// </summary>
+        /// <param name="b1">The first byte of the encoded tag</param>
+        /// <param name="b2">The second byte of the encoded tag</param>
+        /// <param name="b3">The third byte of the encoded tag</param>
+        /// <param name="b4">The fourth byte of the encoded tag</param>
+        public void WriteRawTag(byte b1, byte b2, byte b3, byte b4)
+        {
+            if (buffered + 3 > span.Length)
+            {
+                EnsureMore(4);
+            }
+
+            span[buffered++] = b1;
+            span[buffered++] = b2;
+            span[buffered++] = b3;
+            span[buffered++] = b4;
+        }
+
+        /// <summary>
+        /// Writes the given five-byte tag directly to the stream.
+        /// </summary>
+        /// <param name="b1">The first byte of the encoded tag</param>
+        /// <param name="b2">The second byte of the encoded tag</param>
+        /// <param name="b3">The third byte of the encoded tag</param>
+        /// <param name="b4">The fourth byte of the encoded tag</param>
+        /// <param name="b5">The fifth byte of the encoded tag</param>
+        public void WriteRawTag(byte b1, byte b2, byte b3, byte b4, byte b5)
+        {
+            if (buffered + 4 > span.Length)
+            {
+                EnsureMore(5);
+            }
+
+            span[buffered++] = b1;
+            span[buffered++] = b2;
+            span[buffered++] = b3;
+            span[buffered++] = b4;
+            span[buffered++] = b5;
+        }
+        #endregion
+
+        #region Underlying writing primitives
+        /// <summary>
+        /// Writes a 32 bit value as a varint. The fast route is taken when
+        /// there's enough buffer space left to whizz through without checking
+        /// for each byte; otherwise, we resort to checking size each byte.
+        /// </summary>
+        internal void WriteRawVarint32(uint value)
```

Comment:
the method's comment is copied from CodedOutputStream, but it doesn't look this method implements a fastpath to "whizz through without checking for each byte".

Reply:
I didn't do that because this is already really fast:

```csharp
while (value > 127)
{
    Ensure(1);
    span[buffered++] = (byte)((value & 0x7F) | 0x80);
    value >>= 7;
}

Ensure(1);
span[buffered++] = (byte)value;
```

All `Ensure(1)` will do is a quick if check `if (span.Length < count + buffered)` each iteration, which should be inlined.

I'll remove the comment.